### PR TITLE
Add retries to async_download_binary_file

### PR DESCRIPTION
**Pull request by fepas**
CLOUDBLD-12550

Signed-off-by: Felipe Campos <fdealmei@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- waived Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- n/a OpenAPI schema is updated (if applicable)
- n/a DB schema change has corresponding DB migration (if applicable)
- n/a README updated (if worker configuration changed, or if applicable)

### CODE REVIEW

File: cachito/workers/pkg_managers/general_js.py [hunk 1]
```diff
@@ -104,7 +105,24 @@ async def get_dependencies(
...
+        if current_attempt > 1:
+            package_name = str(params.url).split("/")[-1]
+            log.debug(f"Attempt {current_attempt}/{retry_options.attempts} - {package_name}")
+
+    trace_config = aiohttp.TraceConfig()
+    trace_config.on_request_start.append(on_request_start)
+    retry_options = JitterRetry(attempts=attempts, retry_all_server_errors=True)
+    retry_client = RetryClient(retry_options=retry_options, trace_configs=[trace_config])
+
+    async with retry_client as session:
```
[the remark this answers is not in the document]
Comment (MartinBasti) [reply cachito/workers/pkg_managers/general_js.py:125]: https://github.com/inyutin/aiohttp_retry/blob/master/aiohttp_retry/client.py#L153  it closes just response :(

[the remark this answers is not in the document]
Comment (MartinBasti) [reply cachito/workers/pkg_managers/general_js.py:125]: ah nope, I'm stupid, retry client closes itsellf here https://github.com/inyutin/aiohttp_retry/blob/master/aiohttp_retry/client.py#L389

[the remark this answers is not in the document]
Comment (fepas) [reply cachito/workers/pkg_managers/general_js.py:125]: If one of the tasks fails (it reaches the maximum number of unsuccessful retries (NetworkError)), all the tasks running are canceled. If a task that is in the process of retrying is canceled without closing the retry_client, we receive the following WARNING/ERROR in the logs:

     [2022-12-11 22:56:55,598 aiohttp_retry WARNING client.__del__] Aiohttp retry client was not closed
      [2022-12-11 22:56:55,598 asyncio ERROR base_events.default_exception_handler] Unclosed client session
      client_session: <aiohttp.client.ClientSession object at 0x7f6346725c60>

That's why I'm closing the client before canceling the tasks. If no problem occurs, it closes by itself as you said. :D

[the remark this answers is not in the document]
Comment (MartinBasti) [reply cachito/workers/pkg_managers/general_js.py:125]: I'm puzzled here, where is `session` variable used in this block?

[the remark this answers is not in the document]
Comment (MartinBasti) [reply cachito/workers/pkg_managers/general_js.py:125]: nvm, found it.
I don't know how to do it better, LGTM

**Review by MartinBasti — APPROVED** [review]

**Review by chmeliik — APPROVED** [review]
Looks good :+1: 

This probably does retries a bit differently than our [non-async requests](https://github.com/containerbuildsystem/cachito/blob/master/cachito/workers/requests.py), but since we're using completely different mechanisms for it I think it's close enough

**Comment by chmeliik** [comment]
> Add retry_client to get_dependencies

The title of the PR might make a better commit message, it would also be good to include the Jira ID

And the new code is not covered, but we already gave up on unit-testing aiohttp so ¯\\\_(ツ)\_/¯

File: cachito/workers/pkg_managers/general_js.py [hunk 2]
```diff
@@ -104,7 +105,24 @@ async def get_dependencies(
...
+
+    async def on_request_start(
+        session,
+        trace_config_ctx,
+        params,
+    ) -> None:
+        current_attempt = trace_config_ctx.trace_request_ctx["current_attempt"]
+        if current_attempt > 1:
+            package_name = str(params.url).split("/")[-1]
+            log.debug(f"Attempt {current_attempt}/{retry_options.attempts} - {package_name}")
```
[the remark this answers is not in the document]
Comment (fepas) [reply cachito/workers/pkg_managers/general_js.py:118]: We get the failure that caused the last retry (as before).

**Comment by fepas** [comment]
@chmeliik can you please take a look again? :D

**Review by chmeliik — APPROVED** [review]
LGTM :+1:

**Pull request merged**